Language detection with CLD3

benhalpern opened the pull request:
<!--
     For Work In Progress Pull Requests, please use the Draft PR feature,
     see https://github.blog/2019-02-14-introducing-draft-pull-requests/ for further details.

     For a timely review/response, please avoid force-pushing additional
     commits if your PR already received reviews or comments.

     Before submitting a Pull Request, please ensure you've done the following:
     - 📖 Read the Forem Contributing Guide: https://developers.forem.com/contributing-guide/forem#create-a-pull-request
     - 📖 Read the Forem Code of Conduct: https://github.com/forem/forem/blob/main/CODE_OF_CONDUCT.md
     - 👷‍♀️ Create small PRs. In most cases this will be possible.
     - ✅ Provide tests for your changes.
     - 📝 Use descriptive commit messages.
     - 📗 Update any related documentation and include any relevant screenshots.

     NOTE: Pull Requests from forked repositories will need to be reviewed by
     a Forem Team member before any CI builds will run. Once your PR is approved
     with a `/ci` reply to the PR, it will be allowed to run subsequent builds without
     manual approval.
-->

## What type of PR is this? (check all applicable)

- [ ] Refactor
- [x] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description

This feature adds language detection upon saving, and does so in satisfaction of different conversations we have had about this in the past.

This relies on a well-maintained CLD3 gem (I had previously proposed doing this with a CLD2 gem which was not effectively maintained).

I ran a few benchmarks to ensure that this can be done synchronously.

Saving 1500 articles (while always detecting language due to title changes).

```
Without detecting language: 59.30237799999304ms
With detecting language: 61.59506500000134
```

I also ran a memory profile to see the impact of the gem.

Upon saving a number of articles, always detecting language, here is the breakdown of the memory consumption by gem:

Truncated for readability:

```
allocated memory by gem
-----------------------------------
  74985472  nokogiri-1.15.3-arm64-darwin
  33246836  activesupport-7.0.5.1
  32276551  activerecord-7.0.5.1
  32137718  honeycomb-beeline-2.11.0
    [... truncated]
    206803  openssl
    201576  counter_culture-3.3.1
    190610  cld3-3.5.3 👈
    177872  rouge-3.30.0
    166150  front_matter_parser-1.0.1
    131080  flipper-active_record-0.25.4
    115920  bullet-7.0.7
    112156  zeitwerk-2.6.8
    110132  tzinfo-2.0.6
    102400  logger
     95245  brpoplpush-redis_script-0.1.3
```

As we can see, latency and resource consumption is negligible.

## Related Tickets & Documents

<!--
For pull requests that relate or close an issue, please include them
below.  We like to follow [Github's guidance on linking issues to pull requests](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).

For example having the text: "closes #1234" would connect the current pull
request to issue 1234.  And when we merge the pull request, Github will
automatically close the issue.
-->

- Closes https://github.com/forem/forem/issues/19755

## Added/updated tests?

- [x] Yes
- [ ] No, and this is why: _please replace this line with details on why tests
      have not been included_
- [ ] I need help with writing tests

benhalpern commented on db/migrate/20230713150940_add_language_index_to_articles.rb:
```diff
@@ -0,0 +1,6 @@
+class AddLanguageIndexToArticles < ActiveRecord::Migration[7.0]
+  disable_ddl_transaction!
+  def change
+    add_index :articles, :language, algorithm: :concurrently
```
I proactively added an index here because this is certainly going to be a field which will be queried upon in a variety of ways later.

filleduchaos reviewed (approved):
For QA I did a little bit of mixed language testing using the following as the body_markdown (French title, English body):

```markdown
---
title: Je voudrais quelqu'un m'attende quelque part
published: true
cover_image: https://pigment.github.io/fake-logos/logos/medium/color/5.png
---
I Wish Someone Were Waiting for Me Somewhere is a collection of twelve short stories written by Anna Gavalda.
```

With this I got a `language_outcome.probability` of `0.663392186164856` and `language_outcome.reliable?` was `false`. Adding the line "It was written in French and published in 1999. It is the first book published by Anna Gavalda and it was awarded the Prix Contrepoint in 2001." to the body of the markdown raised the probability to `0.998802900314331` and `language_outcome.reliable?` became `true`.

Weirdly enough the same test with mixed French and Italian gave a high probability (`0.9842222929000854`) with `language_outcome.reliable?` being true right off the bat:

```markdown
---
title: Je voudrais quelqu'un m'attende quelque part
published: true
cover_image: https://pigment.github.io/fake-logos/logos/medium/color/5.png
---
Vorrei Che Da Qualche Parte Ci Fosse Qualcuno Ad Aspettarmi è una raccolta di dodici racconti scritti da Anna Gavalda.
```

jaw6 commented on spec/models/article_spec.rb:
```diff
@@ -1421,6 +1421,47 @@ def foo():
   end
 
+  # rubocop:disable RSpec/AnyInstance
+  describe "#detect_language" do
```
This is probably mostly _fine_ but, at the same time, having spent a _lot_ of the last week thinking about flaky tests specifically and test "discipline" more broadly, there are a few things we could do here to improve this:

* If we introduce a service object as an "adapter" around the `CLD3` interface, the `article` specs can focus on "article things" (like persistence, say)
* With an adapter class, we can write tests specific to testing language (maybe even probability threshold) that don't require hitting any database at all, and maybe don't ever need to build objects more complicated than `String`.
* We don't need to use `update_column` — we'll hit the DB three times for each of the new specs. Instead, if we can need to test the callback, we can use the `article` factory to instantiate a new article with any necessary pre-conditions — we can even use `build` for any tests that aren't specifically testing `update`-related behaviors.

An adapter can be something like:

```ruby
class LanguageDetector
  attr_reader :text
  PROBABILITY_THRESHOLD = 0.5

  def self.call(...)
    new(...).call
  end

  def initialize(text)
    @text = text
  end

  def call(identifier: CLD3::NNetLanguageIdentifier.new(0, 1000))
    language_outcome = identifier.find_language(text)
    return unless language_outcome.probability > PROBABILITY_THRESHOLD && language_outcome.reliable?
    language_outcome.language
  end
end
```

Then `Article#detect_language` can be something like:

```ruby
  def detect_language
    return unless title_changed? || body_markdown_changed?

    self.language = LanguageDetector.call("#{title}. #{body_text}")
  end
```

(The above is not pseudo code, it actually works. I couldn't figure out how to add as a `suggestion` on GitHub, but you can drop it in and it'll work.)

Then we can move the tests that are exploring `CLD3`-specific features (like probability threshold) to specs for the detector specifically and the `article` specs will run faster and be easier to follow. And _none_ of the specs would need to use `any_instance`, which we _really_ shouldn't use.

benhalpern replied:
Thanks for this @jaw6 — definitely no need to settle on "fine" when we're introducing a new thing and have a chance to improve the system overall where we can, so I'll rewrite these parts.

jaw6 reviewed (approved):
The changes I am suggesting aren't strictly _necessary_, but if we aren't under time pressure to merge this, I'd _really_ like to bring the code - especially the specs - up to our standards.

benhalpern commented:
Okay, I've added some tests to test the basic functionality of the library itself @filleduchaos

And @jaw6 I've moved the functionality to its own class, as per your suggestion. Would love a re-review.

jaw6 commented on spec/models/article_spec.rb:
```diff
@@ -1436,27 +1436,18 @@ def foo():
-        ))
+        allow_any_instance_of(Languages::Detection)
+          .to receive(:call)
+          .and_return(:es)
```
You really don't need to mock `any_instance` (and _really **should not**_ if you can at all avoid it). Because `article#detect_language` invokes `Languages::Detect.call`, you can mock the class-level method:

```ruby
  allow(Languages::Detection).to receive(:call).and_return(:es)
```

(As a bonus, you can remove the `#rubocop` bits.)

jaw6 replied:
Also, the way the language mock is layered in here is kind of hard to follow. I'm pretty sure you wind up testing the same behavior multiple times (1433 and 1438 are basically the same test) and not testing the actual interface.

What about something like this approach:

```ruby
describe "#detect_language" do
  let(:detected_language) { :bg }

  before do
    allow(Languages::Detection).to \ 
      receive(:call).and_return(detected_language)
  end
  
  it "should detect language using title and body" do
    [...]
    expect(Languages::Detection).to \
      receive(:call).with("#{title}. #{body_text}")
  end
  
  it "should not call detection when title and body_markdown are unchanged" do
    [...]
    expect(Languages::Detection).not_to \
      have_received(:call)
  end
end
```

jaw6 commented on spec/services/languages/detection_spec.rb:
```diff
@@ -0,0 +1,21 @@
+      expect(subject).to eq(:es)
+    end
+  end
+end
```
These tests do a good job at testing the happy-path. I think the value-add would be to use an `instance_double` to test that `Detection` does "the right thing" (return `nil`) when the probability is low and the result isn't reliable.
